src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,223 @@
+    pExponentialBackoffConfig->maxWaitTime = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_MAX_WAIT_TIME_MILLISECONDS;
+    pExponentialBackoffConfig->retryFactorTime = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_RETRY_TIME_FACTOR_MILLISECONDS;
+    pExponentialBackoffConfig->minTimeToResetRetryState = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_MIN_TIME_TO_RESET_RETRY_STATE_MILLISECONDS;
+    pExponentialBackoffConfig->jitterFactor = DEFAULT_KVS_JITTER_FACTOR_MILLISECONDS;
+
+CleanUp:
+    (*ppPExponentialBackoffConfig) = pExponentialBackoffConfig;
+    LEAVES();
+    return retStatus;
+}
+
+STATUS resetRetryState(PExponentialBackoffState pExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pExponentialBackoffState != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState->currentRetryCount = 0;
+    pExponentialBackoffState->lastRetryWaitTime = 0;
+    pExponentialBackoffState->status = BACKOFF_NOT_STARTED;
+    DLOGD("Resetting Exponential Backoff State");
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS createExponentialBackoffStateWithDefaultConfig(PExponentialBackoffState* ppExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+    PExponentialBackoffConfig pExponentialBackoffConfig = NULL;
+
+    CHK(ppExponentialBackoffState != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState = (PExponentialBackoffState) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffState));
+    CHK(pExponentialBackoffState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    CHK_STATUS(getDefaultExponentialBackOffConfig(&pExponentialBackoffConfig));
+
+    pExponentialBackoffState->pExponentialBackoffConfig = pExponentialBackoffConfig;
+    CHK_STATUS(resetRetryState(pExponentialBackoffState));
+
+    CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppExponentialBackoffState) = pExponentialBackoffState;
+    } else {
+        if (pExponentialBackoffState != NULL) {
+            SAFE_MEMFREE(pExponentialBackoffState->pExponentialBackoffConfig);
+            SAFE_MEMFREE(pExponentialBackoffState);
+        }
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS createExponentialBackoffState(PExponentialBackoffState* ppBackoffState, PExponentialBackoffConfig pBackoffConfig) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+    PExponentialBackoffConfig pExponentialBackoffConfig = NULL;
+
+    CHK(ppBackoffState != NULL, STATUS_NULL_ARG);
+    CHK(pBackoffConfig != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState = (PExponentialBackoffState) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffState));
+    CHK(pExponentialBackoffState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pExponentialBackoffConfig = (PExponentialBackoffConfig) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffConfig));
```

Comment:
Consider making the config as part of the state structure. An extra allocation is an extra allocation which impacts custom allocators on embedded chips.

Reply:
100% agree and thanks for pointing. Makes no sense to do one more allocation. I should not have done this in first place :(